packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts

```diff
@@ -0,0 +1,403 @@
+import { TradeType, validateAndParseAddress } from '@pancakeswap/sdk'
+import {
+  BaseRoute,
+  RouteType,
+  SmartRouter,
+  SmartRouterTrade,
+  StablePool,
+  SwapRouter,
+  getPoolAddress,
+} from '@pancakeswap/smart-router/evm'
+import { Address } from 'viem'
+import { CONTRACT_BALANCE, ROUTER_AS_RECIPIENT, SENDER_AS_RECIPIENT } from '../../utils/constants'
+import { encodeFeeBips } from '../../utils/numbers'
+import { ABIParametersType, CommandType, RoutePlanner } from '../../utils/routerCommands'
+import { Command, RouterTradeType, TradeConfig } from '../Command'
+import { AnyTradeType, PancakeSwapOptions } from '../types'
+
+// Wrapper for pancakeswap router-sdk trade entity to encode swaps for Universal Router
+// also translates trade objects from previous (v2, v3) SDKs
+export class PancakeSwapTrade implements Command {
+  readonly tradeType: RouterTradeType = RouterTradeType.PancakeSwapTrade
+
+  readonly type: TradeType
+
+  constructor(public trade: AnyTradeType, public options: PancakeSwapOptions) {
+    this.type = (this.trade as SmartRouterTrade<TradeType>).tradeType
+  }
+
+  encode(planner: RoutePlanner, _config: TradeConfig): void {
+    let payerIsUser = true
+    const trades = !Array.isArray(this.trade) ? [this.trade] : this.trade
+    const numberOfTrades = trades.reduce((numOfTrades, trade) => numOfTrades + trade.routes.length, 0)
+
+    const sampleTrade = trades[0]
+    // If the input currency is the native currency, we need to wrap it with the router as the recipient
+    if (sampleTrade.inputAmount.currency.isNative) {
+      // TODO: optimize if only one v2 pool we can directly send this to the pool
+      planner.addCommand(CommandType.WRAP_ETH, [
+        ROUTER_AS_RECIPIENT,
+        BigInt(
+          SmartRouter.maximumAmountIn(
+            sampleTrade,
+            this.options.slippageTolerance,
+            sampleTrade.inputAmount,
+          ).quotient.toString(),
+        ),
+      ])
+      // since WETH is now owned by the router, the router pays for inputs
+      payerIsUser = false
+    }
+    // The overall recipient at the end of the trade, SENDER_AS_RECIPIENT uses the msg.sender
+    this.options.recipient = this.options.recipient ?? SENDER_AS_RECIPIENT
+
+    const inputIsNative = sampleTrade.inputAmount.currency.isNative
+    const outputIsNative = sampleTrade.outputAmount.currency.isNative
+    const performAggregatedSlippageCheck = sampleTrade.tradeType === TradeType.EXACT_INPUT && numberOfTrades > 2
+    const routerMustCustody = outputIsNative || !!this.options.fee || performAggregatedSlippageCheck
+
+    for (const trade of trades) {
+      if (trade.routes.length === 1 && trade.routes[0].type === RouteType.V2) {
```

Comment:
If all the routes are v2 then we should add v2 swap

Reply:
We need to loop the routes here? Because a single route is considered a valid swap